File: ccwap/server/models/analytics.py
"""Pydantic models for deep analytics API."""

from typing import List, Optional, Dict
from pydantic import BaseModel


class ThinkingAnalysis(BaseModel):
    """Extended thinking analysis."""
    total_thinking_chars: int = 0
    avg_thinking_per_turn: float = 0.0
    turns_with_thinking: int = 0
    total_turns: int = 0
    thinking_rate: float = 0.0
    by_model: List[Dict] = []


class TruncationAnalysis(BaseModel):
    """Truncation/stop reason breakdown."""
    total_turns: int = 0
    by_stop_reason: List[Dict] = []


class SidechainAnalysis(BaseModel):
    """Sidechain/branching analysis."""
    total_sidechains: int = 0
    sidechain_rate: float = 0.0
    by_project: List[Dict] = []


class CacheTierAnalysis(BaseModel):
    """Ephemeral cache tier analysis."""
    ephemeral_5m_tokens: int = 0
    ephemeral_1h_tokens: int = 0
    standard_cache_tokens: int = 0
    by_date: List[Dict] = []


class BranchAnalytics(BaseModel):
    """Branch-aware analytics."""
    branches: List[Dict] = []


class VersionImpact(BaseModel):
    """CC version impact analysis."""
    versions: List[Dict] = []


class SkillsAgents(BaseModel):
    """Skills and agents analysis."""
    total_agent_spawns: int = 0
    total_skill_invocations: int = 0
    agent_cost: float = 0.0
    by_date: List[Dict] = []


class AnalyticsResponse(BaseModel):
    """Complete deep analytics data."""
    thinking: ThinkingAnalysis
    truncation: TruncationAnalysis
    sidechains: SidechainAnalysis
    cache_tiers: CacheTierAnalysis
    branches: BranchAnalytics
    versions: VersionImpact
    skills_agents: SkillsAgents

File: ccwap/frontend/src/api/analytics.ts
import { useQuery } from '@tanstack/react-query'
import { apiFetch, buildQuery } from './client'
import { analyticsKeys } from './keys'
import type { DateRange } from '@/hooks/useDateRange'

export interface ThinkingAnalysis {
  total_thinking_chars: number
  avg_thinking_per_turn: number
  turns_with_thinking: number
  total_turns: number
  thinking_rate: number
  by_model: Record<string, unknown>[]
}

export interface TruncationAnalysis {
  total_turns: number
  by_stop_reason: Record<string, unknown>[]
}

export interface SidechainAnalysis {
  total_sidechains: number
  sidechain_rate: number
  by_project: Record<string, unknown>[]
}

export interface CacheTierAnalysis {
  ephemeral_5m_tokens: number
  ephemeral_1h_tokens: number
  standard_cache_tokens: number
  by_date: Record<string, unknown>[]
}

export interface BranchAnalytics {
  branches: Record<string, unknown>[]
}

export interface VersionImpact {
  versions: Record<string, unknown>[]
}

export interface SkillsAgents {
  total_agent_spawns: number
  total_skill_invocations: number
  agent_cost: number
  by_date: Record<string, unknown>[]
}

export interface AnalyticsData {
  thinking: ThinkingAnalysis
  truncation: TruncationAnalysis
  sidechains: SidechainAnalysis
  cache_tiers: CacheTierAnalysis
  branches: BranchAnalytics
  versions: VersionImpact
  skills_agents: SkillsAgents
}

export function useAnalytics(dateRange: DateRange) {
  return useQuery({
    queryKey: analyticsKeys.data(dateRange),
    queryFn: () => apiFetch<AnalyticsData>(
      `/analytics${buildQuery({ from: dateRange.from, to: dateRange.to })}`
    ),
  })
}

export interface ThinkingTrendPoint {
  date: string
  model: string
  thinking_chars: number
}

export interface CacheTrendPoint {
  date: string
  ephemeral_5m: number
  ephemeral_1h: number
  standard_cache: number
}

export function useThinkingTrend(dateRange: DateRange) {
  return useQuery({
    queryKey: analyticsKeys.thinkingTrend(dateRange),
    queryFn: () => apiFetch<ThinkingTrendPoint[]>(
      `/analytics/thinking-trend${buildQuery({ from: dateRange.from, to: dateRange.to })}`
    ),
  })
}

export function useCacheTrend(dateRange: DateRange) {
  return useQuery({
    queryKey: analyticsKeys.cacheTrend(dateRange),
    queryFn: () => apiFetch<CacheTrendPoint[]>(
      `/analytics/cache-trend${buildQuery({ from: dateRange.from, to: dateRange.to })}`
    ),
  })
}
